DescriptorMatcher::checkMasks fails on Nx1 masks

### Please state the information for your system
- OpenCV version: 3.x
- Host OS: Linux (Ubuntu 14.04)

### In which part of the OpenCV library you got the issue?
- feature2d
- matcher

### Expected behaviour
Running knnMatch or radiusMatch with a Nx1 mask pass on the assertion inside DescriptorMatcher::checkMasks method.

### Actual behaviour
Running fails with the following error
> 
matchers.cpp:617: error: (-215) masks[i].rows == queryDescriptorsCount && (masks[i].cols == rows || masks[i].cols == rows) && masks[i].type() == CV_8UC1 in function checkMasks

### Additional description
The assertion fails becaus the second condition of the assertion only compares the number of columns of the masks.

### Code example to reproduce the issue / Steps to reproduce the issue
```c++
#include <iostream>

#include <string>
#include <vector>

#include <opencv2/highgui.hpp>
#include <opencv2/videoio.hpp>

#include <opencv2/features2d.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/video/video.hpp>

int main(int argc, char** argv) {

    std::string folder("");

    std::vector<cv::Mat> images;
    images.push_back(cv::imread("/home/anishitani/workspace/visionws/src/tutorial/data/fountain/0000.jpg",cv::IMREAD_GRAYSCALE));
    images.push_back(cv::imread("/home/anishitani/workspace/visionws/src/tutorial/data/fountain/0001.jpg",cv::IMREAD_GRAYSCALE));

    /** Feature detection init **/
    std::vector<std::vector<cv::KeyPoint> > fkpts(2);

    cv::Ptr<cv::FeatureDetector> detector = cv::FastFeatureDetector::create();
    detector->detect(images[0],fkpts[0]);
    detector->detect(images[1],fkpts[1]);
    /** Feature detection end **/

    /** Feature matching init **/
    std::vector<std::vector<cv::Point2f> > fpts(2);
    cv::KeyPoint::convert(fkpts[0],fpts[0]);
    cv::KeyPoint::convert(fkpts[1],fpts[1]);
    // Calculate the optical flow field:
    // how each left_point moved across the 2 images
    std::vector<uchar> vstatus;
    std::vector<float> verror;
    cv::calcOpticalFlowPyrLK(images[0], images[1], fpts[0], fpts[1], vstatus, verror);
    /** Feature matching end **/

    cv::Mat out;
    cv::cvtColor(images[0],out,cv::COLOR_GRAY2BGR);
    std::vector<cv::Point2f> flowpts;
    for(int i=0 ; i<fpts[0].size() ; i++){
        if(vstatus[i] && verror[i]<12.0){
            flowpts.push_back(fpts[1][i]);
        } else {
            flowpts.push_back(cv::Point2f());
            vstatus[i] = 0;
        }
    }

    cv::BFMatcher matcher;
    std::vector<std::vector<cv::DMatch> > nn;
    cv::Mat mask;
    cv::Mat(vstatus).convertTo(mask,CV_8UC1);

    matcher.radiusMatch( cv::Mat(fpts[1]).reshape(1,fpts[1].size()),
            cv::Mat(flowpts).reshape(1,flowpts.size()),
            nn,2.0f, mask);

    cv::namedWindow("features",cv::WINDOW_KEEPRATIO);
    cv::imshow("features",out);

    for(char k = cv::waitKey() ; k!='q' ; k = cv::waitKey());

    return 0;
}

```

--- a/modules/features2d/src/matchers.cpp
+++ b/modules/features2d/src/matchers.cpp
@@ -613,8 +613,7 @@ void DescriptorMatcher::checkMasks( InputArrayOfArrays _masks, int queryDescript
             {
                 int rows = trainDescCollection[i].empty() ? utrainDescCollection[i].rows : trainDescCollection[i].rows;
                     CV_Assert( masks[i].rows == queryDescriptorsCount &&
-                        (masks[i].cols == rows || masks[i].cols == rows) &&
-                        masks[i].type() == CV_8UC1 );
+                        masks[i].cols == rows && masks[i].type() == CV_8UC1);
             }
         }
     }
@@ -1357,5 +1356,4 @@ void FlannBasedMatcher::radiusMatchImpl( InputArray _queryDescriptors, std::vect
 
     convertToDMatches( mergedDescriptors, indices, dists, matches );
 }
-
 }

